src/shims/sync.rs

```diff
@@ -454,135 +456,325 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     fn pthread_rwlock_unlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
-        let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
-        let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
-        let writer_blockset = rwlock_get_or_create_writer_blockset(this, rwlock_op)?;
-        if let Some(new_readers) = readers.checked_sub(1) {
-            assert_eq!(writers, 0);
-            rwlock_set_readers(this, rwlock_op, Scalar::from_u32(new_readers))?;
-            if new_readers == 0 {
-                if let Some(_writer) = this.unblock_some_thread(writer_blockset)? {
-                    rwlock_set_writers(this, rwlock_op, Scalar::from_u32(1))?;
+        let id = rwlock_get_or_create_id(this, rwlock_op)?;
+        let active_thread = this.get_active_thread()?;
+
+        if this.rwlock_reader_remove(id, active_thread) {
+            // The thread was a reader.
+            if this.rwlock_is_locked(id) {
+                // No more readers owning the lock. Give it to a writer if there
+                // is any.
+                if let Some(writer) = this.rwlock_dequeue_writer(id) {
+                    this.unblock_thread(writer)?;
+                    this.rwlock_writer_set(id, writer);
                 }
             }
             Ok(0)
-        } else if writers != 0 {
-            let reader_blockset = rwlock_get_or_create_reader_blockset(this, rwlock_op)?;
+        } else if Some(active_thread) == this.rwlock_writer_remove(id) {
+            // The thread was a writer.
+            //
             // We are prioritizing writers here against the readers. As a
             // result, not only readers can starve writers, but also writers can
             // starve readers.
-            if let Some(_writer) = this.unblock_some_thread(writer_blockset)? {
-                assert_eq!(writers, 1);
+            if let Some(writer) = this.rwlock_dequeue_writer(id) {
+                // Give the lock to another writer.
+                this.unblock_thread(writer)?;
+                this.rwlock_writer_set(id, writer);
             } else {
-                rwlock_set_writers(this, rwlock_op, Scalar::from_u32(0))?;
-                let mut readers = 0;
-                while let Some(_reader) = this.unblock_some_thread(reader_blockset)? {
-                    readers += 1;
+                // Give the lock to all readers.
+                while let Some(reader) = this.rwlock_dequeue_reader(id) {
+                    this.unblock_thread(reader)?;
+                    this.rwlock_reader_add(id, reader);
                 }
-                rwlock_set_readers(this, rwlock_op, Scalar::from_u32(readers))?
             }
             Ok(0)
         } else {
-            throw_ub_format!("unlocked an rwlock that was not locked");
+            throw_ub_format!("unlocked an rwlock that was not locked by the active thread");
```

Comment:
So we can now detect when we try to release another thread's rwlock... is there a test for that?